File: code/ch03/3.6.1.superpalindromesInRange.py
import math


class Solution:
    def superpalindromesInRange(self, L: str, R: str) -> int:
        cnt = 0

        def validPalindrome(s: str) -> bool:
            l = 0
            r = len(s) - 1
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        for i in range(math.floor(int(L) ** 0.5), math.ceil(int(R) ** 0.5)):
            if validPalindrome(str(i)) and validPalindrome(str(i ** 2)):
                cnt += 1
        return cnt

File: code/ch03/3.6.1.superpalindromesInRange.java
import java.lang.Math;

public class Solution {
  public boolean validPalindrome(String s) {
    int l = 0, r = s.length() - 1;
    while (l < r) {
      if (s.charAt(l) != s.charAt(r)) return false;
      l += 1;
      r -= 1;
    }
    return true;
  }
  public int superpalindromesInRange(String L, String R) {
    int cnt = 0;
    long lbound = (long)Math.floor(Math.sqrt(Long.parseLong(L)));
    long rbound = (long)Math.ceil(Math.sqrt(Long.parseLong(R)));
    for (long i = lbound; i < rbound; i += (long)1)
      if (validPalindrome(String.valueOf(i)) &&
          validPalindrome(String.valueOf(i * i)))
        cnt = cnt + 1;
    return cnt;
  }
}
